src/Microsoft.AspNetCore.Sockets/Transports/WebSocketsTransport.cs

```diff
@@ -140,43 +152,43 @@ private void LogFrame(string action, WebSocketFrame frame)
 
         private async Task StartSending(IWebSocketConnection ws)
         {
-            try
+            while (!_connection.Input.Completion.IsCompleted)
             {
-                while (true)
+                // Get a frame from the application
+                try
                 {
-                    var result = await _channel.Output.ReadAsync();
-                    var buffer = result.Buffer;
-
-                    try
+                    using (var message = await _connection.Input.ReadAsync())
                     {
-                        if (buffer.IsEmpty && result.IsCompleted)
+                        if (message.Payload.Buffer.Length > 0)
                         {
-                            break;
+                            try
+                            {
+                                var opcode = message.MessageFormat == Format.Binary ?
+                                    WebSocketOpcode.Binary :
+                                    WebSocketOpcode.Text;
+
+                                var frame = new WebSocketFrame(
+                                    endOfMessage: message.EndOfMessage,
+                                    opcode: _lastFrameIncomplete ? WebSocketOpcode.Continuation : opcode,
+                                    payload: message.Payload.Buffer);
+
+                                _lastFrameIncomplete = !message.EndOfMessage;
+
+                                LogFrame("Sending", frame);
+                                await ws.SendAsync(frame);
+                            }
+                            catch (Exception ex)
+                            {
+                                _logger.LogError("Error writing frame to output: {0}", ex);
+                                break;
+                            }
                         }
-
-                        // Send the buffer in a frame
-                        var frame = new WebSocketFrame(
-                            endOfMessage: true,
-                            opcode: _opcode,
-                            payload: buffer);
-                        LogFrame("Sending", frame);
-                        await ws.SendAsync(frame);
-                    }
-                    catch (Exception ex)
-                    {
-                        _logger.LogError("Error writing frame to output: {0}", ex);
-                        break;
-                    }
-                    finally
-                    {
-                        _channel.Output.Advance(buffer.End);
                     }
                 }
-            }
-            finally
-            {
-                // No longer reading from the channel
-                _channel.Output.CompleteReader();
+                catch (Exception ex) when (ex.GetType().IsNested && ex.GetType().DeclaringType == typeof(Channel))
```

Comment:
Fixed in https://github.com/dotnet/corefxlab/pull/1097, just waiting for it to propagate